File: pdns-backend.rb
#! /usr/bin/ruby
# -*- coding: utf-8 -*-

require 'json'
require 'pp'
require 'syslog'


class LIST
  attr_accessor :ip, :weight, :ttl

  def initialize(ip, weight, ttl)
    @ip = ip
    @weight = weight
    @ttl = ttl
  end
end


def random_choice(dup)
  dup.sort{|a,b| a.weight <=> b.weight}
  weight_sum = 0
  dup.each do |d|
    weight_sum +=  d.weight
  end
  dup.each do |d|
    d.weight /= weight_sum
  end
  r = rand()
  sum = 0
  dup.each_with_index do |d, n|
    sum += d.weight
    return n if r < sum
  end
  return dup.length - 1
end


def sort(ip,select_num)
  lis1 = Array.new
  lis2 = Array.new
  cnt = 0

  # ipアドレスと重みを配列に格納
  ip.each_with_index do |i, n|
 #   s = i.split(" ")
    lis1[n] = LIST.new(i["ip"], i["weight"].to_f, i["ttl"].to_i) 
  end
  
  # 重み付けしてソート
  loop{
    # すべてのIPの重みが0の場合
    if !checkweight(lis1) then
      if lis2.length == 0 then
        return lis1
      end
      break
    end
    if lis1.length == 1 && cnt == 0 then
      lis2.push(lis1[0])
        break
    end
    if lis1.length == 1 then
      #lis2.push(lis1[0])
#        break
    end
    break if cnt >= select_num.to_i # いくつIPアドレスを返すか
    index = random_choice(lis1)
    lis2.push(lis1[index])
    lis1.delete_at(index)
    cnt += 1
  }
  return lis2
end

def checkhash(qname, qtype)

  qtype = "A" if qtype == "ANY"
 
  return false if @@config["domain"] != qname
  return false if @@config["type"] != qtype.downcase
  return true
  
end

def checkweight(lis)

  lis.each do |i|
    if i.weight != 0 then
      return true
    end
  end
  return false
end


def minimum_ttl(list)
  ttl = 1000
  list.each do |i|
    if i.ttl < ttl
      ttl = i.ttl
    end
  end
  
  return ttl
  
end


$stdout.sync = true
$syslog = Syslog.open(__FILE__)
END { $syslog.close }

line = gets
line.strip!

unless line == "HELO\t1"
  puts "FAIL"
  #$syslog.err "Recevied '#{line}'"
  gets
  exit
end

puts "OK Sample backend firing up\t"

while gets
  #$syslog.info "#{$$} Received: #{$_}"
  $_.strip!
  arr = $_.split(/\t/)

  if (arr.length < 6)
    puts "LOG\tPowerDNS sent unparseable line"
    puts "FAIL"
    next
  end
  type, qname, qclass, qtype, id, ip = arr
  #$syslog.info "#{type}, #{qname}, #{qclass}, #{qtype}, #{id}"

  File.open("/home/vagrant/backend/wrr-config.json") do |file|
    @@config = JSON.load(file)
  end
  
  if checkhash(qname, qtype)
    if ["A", "ANY"].any? {|i| qtype == i } 
      arr = sort(@@config["record"], @@config["num"])
      ttl = minimum_ttl(arr)
      #$syslog.info "#{$$} Sent A records"
      arr.each do |i|
        #$syslog.info  ["send : ", "DATA", qname, qclass, "A", i.ttl, 1, i.ip].join(" ")
        puts ["DATA", qname, qclass, "A", ttl, 1, i.ip].join("\t")
      end
    end
  end

  #$syslog.info "#{$$} End of data"
  puts "END"

end

File: pdns-backend.go
package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"log"
	"math/rand"
	"os"
	"sort"
	"strconv"
	"strings"
	"time"
)

type PipeBackendStruct struct {
	Domain  string         `json:"domain"`
	Type    string         `json:"type"`
	Num     int            `json:"num"`
	Records []RecordStruct `json:"record"`
}

type RecordStruct struct {
	IP     string `json:"ip"`
	Weight int    `json:"weight"`
	TTL    int    `json:"ttl"`
}

func main() {
	rand.Seed(time.Now().UnixNano())

	backend := PipeBackendStruct{}

	scanner := bufio.NewScanner(os.Stdin)
	//output("output", "Start Backend Program")

	scanner.Scan()
	s := strings.TrimRight(scanner.Text(), "\n")
	if s != "HELO\t1" {
		fmt.Println("FAIL")
		os.Exit(0)
	}
	fmt.Println("OK Sample Backend firing up\t")

	for scanner.Scan() {
		strs := strings.TrimRight(scanner.Text(), "\n")
		str := strings.Split(strs, "\t")
		if len(str) < 6 {
			fmt.Println("LOG\tPowerDNS sent unparseable line")
			continue
		}
		bytes, err := ioutil.ReadFile("/home/vagrant/backend/config.json")
		if err != nil {
			outputError(err)
			return
		}
		err = json.Unmarshal(bytes, &backend)
		if err != nil {
			outputError(err)
			return
		}
		switch str[3] {
		case "A", "ANY":
			if str[1] == backend.Domain {
				ips, ttl := selectIPs(backend)
				for _, i := range ips {
					if i == "" {
						break
					}
					fmt.Println("DATA\t" + str[1] + "\t" + str[2] + "\tA\t" + strconv.Itoa(ttl) + "\t" + "1\t" + i)
					//output("DATA\t" + str[1] + "\t" + str[2] + "\tA\t" + strconv.Itoa(ttl) + "\t" + "-1\t" + i)
				}
				go outputTTL(ttl)

			} else {
				output(str)
			}
		case "SOA":
			if str[1] == "sai.test" {
				fmt.Println("DATA\t" + str[1] + "\t" + str[2] + "\tSOA\t60\t-1\tsai.test. root.sai.test 2017112601 60 60 60 60")
				//output("DATA\t" + str[1] + "\t" + str[2] + "\tSOA\t60\t-1\tsai.test. root.sai.test 2017112601 60 60 60 60")
			}
		default:
			//output(str)
		}
		fmt.Println("END")
	}
}

func selectIPs(backend PipeBackendStruct) ([]string, int) {

	ttl := 10000000
	ips := make([]string, backend.Num)

	if len(backend.Records) == 1 {
		ips[0] = backend.Records[0].IP
		return ips, backend.Records[0].TTL
	}
	// すべての重みが0のときの処理
	if checkWeight(backend.Records) {
		for i := 0; i < backend.Num; i++ {
			ips[i] = backend.Records[i].IP
			if ttl > backend.Records[i].TTL {
				ttl = backend.Records[i].TTL
			}
		}
		return ips, ttl
	}

	//records := make([]RecordStruct, len(backend.Records))
	records := copyStruct(backend.Records)

	for k := 0; k < backend.Num && len(records) > 0; k++ {
		weight := make([]int, len(records))
		for i, v := range records {
			weight[i] = v.Weight
		}

		boundaries := make([]int, len(records)+1)
		for i := 1; i < len(boundaries); i++ {
			boundaries[i] = boundaries[i-1] + weight[i-1]
		}

		boundaryLast := int(boundaries[len(boundaries)-1])
		x := rand.Intn(boundaryLast) + 1
		idx := sort.SearchInts(boundaries, int(x)) - 1

		ips[k] = records[idx].IP
		if ttl > records[idx].TTL {
			ttl = records[idx].TTL
		}
		records = pop(records, idx)
	}
	return ips, ttl
}

func copyStruct(slice []RecordStruct) []RecordStruct {

	records := []RecordStruct{}

	for _, v := range slice {
		if v.Weight != 0 {
			records = append(records, v)
		}
	}
	return records

}

func pop(slice []RecordStruct, index int) []RecordStruct {

	result := []RecordStruct{}
	for i, v := range slice {
		if i != index {
			result = append(result, v)
		}
	}
	return result
}

func checkWeight(slice []RecordStruct) bool {

	for _, v := range slice {
		if v.Weight != 0 {
			return false
		}
	}

	return true

}

func outputError(err error) {

	f, err := os.OpenFile("/home/vagrant/backend/error.txt", os.O_RDWR|os.O_CREATE|os.O_APPEND, 0644)
	defer f.Close()
	if err != nil {
		return
	}
	log.SetOutput(f)
	log.Println(err)
}

func output(str interface{}) {

	f, err := os.OpenFile("/home/vagrant/backend/output.txt", os.O_RDWR|os.O_CREATE|os.O_APPEND, 0644)
	defer f.Close()
	if err != nil {
		return
	}
	log.SetOutput(f)
	log.Print(fmt.Sprintln(str))
}

func outputTTL(str interface{}) {

	now := time.Now().Format("15:04:05")

	f, err := os.OpenFile("/home/vagrant/backend/ttl.csv", os.O_RDWR|os.O_CREATE|os.O_APPEND, 0644)
	defer f.Close()
	if err != nil {
		outputError(err)
		return
	}

	_, err = f.WriteString(now + "," + fmt.Sprintln(str))
	if err != nil {
		outputError(err)
	}
	return
}
